### ISSUE TITLE
Property setter not followed when checking assignment type compatibility

### ISSUE THREAD

**Issue by odorikakeru** [python/mypy#16411]
**Bug Report**

In cases where a class property returns one type, but the setter for that property accepts a union of types: when evaluating an assignment to that property, mypy will only check the return value of the initial property method (with the `@property` decorator and will not check the setter method. This leads to false positives when trying to assign to the property using a type that is accepted by the setter method, but different from the return value of the getter.

**To Reproduce**

mypy Playground Gist: https://mypy-play.net/?mypy=latest&python=3.11&gist=699b298675e7029e8a8109d221b6e24f

```python
from typing import Union


class Stringifier:
    
    def __init__(self) -> None:
        self._mystr: str = ''
    
    @property
    def mystr(self) -> str:
        return self._mystr
        
    @mystr.setter
    def mystr(self, value: Union[str, int, float]) -> None:
        self._mystr = str(value)
        

if __name__ == '__main__':
    myobj = Stringifier()
    myint: int = 5
    
    myobj.mystr = myint
    
    print(myobj.mystr)  # '5'
```

**Expected Behavior**

prints the string '5' to the screen

**Actual Behavior**

`error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]`

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: -
- Mypy configuration options from `mypy.ini` (and other config files): -
- Python version used: 3.11.3

**Comment by erictraut** [comment]
This is related to (and perhaps a duplicate of) https://github.com/python/mypy/issues/3004.

**Comment by ilevkivskyi** [comment]
Duplicate of #3004
